source/Mlos.Python/mlos/Optimizers/ParetoFrontier.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+import pandas as pd
+from mlos.Optimizers.OptimizationProblem import OptimizationProblem
+
+
+class ParetoFrontier:
+    """Maintains a set of non-dominated configurations.
+
+    This class will have several pieces of functionality:
+        1. It will be able to construct and maintain a pareto frontier from a set of observations for one or more objectives.
+        2. It will be able to update the frontier upon receiving a new observation.
+        3. It will be able to decide whether any given point is dominated or not (needed for Monte Carlo utility functions).
+
+    Each point will be characterized by:
+        1. Configuration parameters
+        2. Objective function values
+        3. Possibly context values
+
+    A point belongs to a pareto frontier if it is not dominated by any other point. So if two points have the exact same values
+    for all objectives (but possibly different configurations), we will consider both of them to be pareto efficient.
+
+
+    """
+
+    def __init__(
+            self,
+            optimization_problem: OptimizationProblem,
+            features_df: pd.DataFrame,
+            objectives_df: pd.DataFrame
+    ):
+        assert len(features_df.index) == len(objectives_df.index)
+        if features_df is not None:
+            assert all(column in optimization_problem.feature_space.dimension_names for column in features_df.columns)
+
+        if objectives_df is not None:
+            assert all(column in optimization_problem.objective_space.dimension_names for column in objectives_df.columns)
+
+        self.optimization_problem = optimization_problem
+        self.features_df = features_df
+        self.objectives_df = objectives_df
+
+    @staticmethod
+    def compute_pareto(optimization_problem: OptimizationProblem, objectives_df: pd.DataFrame) -> pd.DataFrame:
+        """Computes a pareto frontier for the given objectives_df.
+
+        We do this by consecutively removing dominated points from objectives_df until none are left.
+
+
+        :param optimization_problem:
+        :param objectives_df:
+        :return:
+        """
+
+        assert all(column in optimization_problem.objective_space.dimension_names for column in objectives_df.columns)
+
+        # Let's copy it, since we are going to mess it up.
+        #
+        pareto_df = objectives_df.copy(deep=True)
+
+        # First, let's turn it into a maximization problem, by flipping the sign of all objectives that are to be minimized.
+        #
+        for objective in optimization_problem.objectives:
+            if objective.minimize:
+                pareto_df[objective.name] = -pareto_df[objective.name]
+
+        # By presorting we guarantee, that all dominated points are below the currently considered point.

```

Comment:
That's only true if no objective value appears twice, right?

Reply:
At the time you made this comment, the sort was (erroneously) using only the first objective. Nice catch :)

I believe that sorting on all objectives fixes the problem. I got it to produce this:

![image](https://user-images.githubusercontent.com/51930563/99687470-b0649a00-2a52-11eb-8194-c770b5b0ab8c.png)


@edcthayer: during our conversation yesterday we considered the green points:

On the one hand they are on the frontier line, and we will know the corresponding configuration.

On the other hand they are dominated in that we could improve one of the objectives, without sacrificing another. Excluding them, would produce a smaller dataset for the user, though I really expect this to be a rare occurrence as most objectives will likely be continuous and a probability of getting the exact same float value twice are slim.

So, should we include these or not?

Reply:
Talked with Ed. We decided that including these points is better for 3 reasons:
1. If they were jittered only a bit to the north-east we would want to include them.
2. The code to compute this is more efficient.
3. It's a very rare situation in real life, so if the user doesn't appreciate these points, chances are that they will never see them :)

Reply:
Yes, I agree this is an edge case for sure, though we should make sure to handle it semi-reasonable :)

I wonder if these are weak pareto points? https://en.wikipedia.org/wiki/Pareto_efficiency
I need to reread the wikipedia article. Maybe adding a comment to the test and the class might be good?

Reply:
Actually, I think sorting only on the first dimension would work again, but I'll leave it in place in case we ever want to exclude the "maybe" points.